File: assign3/scala/src/pizza/Pizza.scala
package pizza

package object PizzaDSL {
	class Pizza {
		var layer = List[String]()
		
		def spread(component : String*) = {
			layer = layer ++ component
			this
		}
		
		def toppings(component : String*) = {
			layer = layer ++ component
			this
		}
		
		override def toString = {
			s"Pizza with ${if(layer == List()) "no layer" else s"${layer.length} layers: ${layer.mkString(", ")}"}"
		}
	}
	
	class Create {
		def a(pizza: Pizza) = new Pizza()
	}

	val create = new Create()
	val pizza = new Pizza()
	val sauce = "sauce"
	val cheese = "cheese"
	val onions = "onions"
	val meat = "meat"
}

File: assign3/ruby/src/pizza.rb
class Pizza

	attr_accessor :name, :layer
	
	def initialize(name)
		@name = name
		@layer = []
	end
 
	def spread(*component)
		component.map { |comp| @layer << comp}
	end
	
	def toppings(*component)
		component.map { |comp| @layer << comp}
	end
	
	def to_s
		"#{name} with #{@layer.count != 0 ? "#{@layer.count} layers: #{@layer.join(', ')}" : 'no layer'}"
	end
	
	def method_missing(name, *args)
		name.to_sym
	end
end

def pizza
	"Pizza"
end

def create(name, &block)
	pizza = Pizza.new(name)
	pizza.instance_eval(&block) unless block.nil?
	pizza
end